source/PlayerInfo.cpp

```diff
@@ -2639,6 +2639,182 @@ void PlayerInfo::RegisterDerivedConditions()
 		return retVal;
 	});
 
+	// The numver of ships the player has of the given category anywhere in their fleet.
+	auto &&shipTypesAllProvider = conditions.GetProviderPrefixed("ships (all): ");
+	shipTypesAllProvider.SetGetFunction([this](const string &name) -> int64_t
+	{
+		int64_t retVal = 0;
+		for(const shared_ptr<Ship> &ship : ships)
+			if(!ship->IsDestroyed() && name == "ships (all): " + ship->Attributes().Category())
+				++retVal;
+		return retVal;
+	});
+
+	// The number of ships the player has of the given model active and present.
+	auto &&shipModelProvider = conditions.GetProviderPrefixed("ship model: ");
+	shipModelProvider.SetGetFunction([this](const string &name) -> int64_t
+	{
+		int64_t retVal = 0;
+		for(const shared_ptr<Ship> &ship : ships)
+			if(!ship->IsParked() && !ship->IsDisabled() && ship->GetSystem() == system
```

Comment:
#6628 mentions how the `"ships: <category>"` call doesn't properly check docked fighters. This is as a result of how docked fighters have their system set to nullptr, therefore causing the `ship->GetSystem() == system` call to return false.

https://github.com/endless-sky/endless-sky/blob/7f708d010a215fb81fe80fc732445d882c53f3f5/source/Ship.cpp#L3586-L3591

This same skipping of fighters would affect many of these new conditions. In order to properly capture carried fighters, would we want to do make the following check instead?
```c++
const System *shipSystem = ship->GetSystem() ? ship->GetSystem() : ship->GetParent()->GetSystem();
if(shipSystem == system)
	...
```

Reply:
This fix looks fine, but I have 2 remarks:
- A ship might both not have a system, and not a parent. Not sure when, but let's just be safe and check if GetParent() returns a pointer before using it.
- The function to get the ships system (and the parents one if the ship is docked) seems best placed within Ship.cpp, with a function-name like `Ship::GetSystemFull` or `Ship::GetSystemComplete` or `Ship::GetSystemRecursive`. By placing this in ship we can also re-use it when we need it on other places.